File: src/main/java/com/haozileung/infra/web/BaseServlet.java
/**
 * 
 */
package com.haozileung.infra.web;

import java.io.IOException;
import java.io.PrintWriter;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.beetl.ext.servlet.ServletGroupTemplate;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.alibaba.fastjson.JSON;
import com.google.common.base.Strings;

/**
 * @author Haozi
 *
 */
public abstract class BaseServlet extends HttpServlet {

	@Override
	public void init() throws ServletException {
		if (Initializer.getInjector() != null) {
			Initializer.getInjector().injectMembers(this);
		}
	}

	/**
	 * 
	 */
	private static final long serialVersionUID = 7514613990641762954L;
	private final static Logger logger = LoggerFactory.getLogger(BaseServlet.class);

	protected void renderView(int code, String view, HttpServletRequest req, HttpServletResponse resp) {
		resp.setContentType("text/html;charset=utf-8");
		resp.setStatus(code);
		if (!Strings.isNullOrEmpty(view)) {
			try {
				ServletGroupTemplate.instance().render(view, req, resp);
			} catch (Exception e) {
				logger.error(e.getMessage(), e);
				try {
					resp.sendError(HttpServletResponse.SC_NOT_FOUND);
				} catch (IOException e1) {
					logger.error(e1.getMessage(), e1);
				}
			}
		}
	}

	protected void renderJSON(int code, Object data, HttpServletResponse resp) {
		resp.setContentType("application/json;charset=utf-8");
		resp.setStatus(code);
		try {
			PrintWriter pr = resp.getWriter();
			pr.print(JSON.toJSONString(data));
			pr.flush();
		} catch (IOException e) {
			logger.error(e.getMessage(), e);
		}
	}

	protected void render(Object result, HttpServletRequest req, HttpServletResponse resp) {
		if (result != null) {
			if (result instanceof String) {
				String s = (String) result;
				if (s.startsWith("redirect:")) {
					s = s.replace("redirect:", "");
					try {
						resp.sendRedirect(s);
					} catch (IOException e) {
						logger.error(e.getMessage(), e);
					}
				}
				renderView(resp.getStatus(), s, req, resp);
			} else {
				renderJSON(resp.getStatus(), result, resp);
			}
		}
	}

	@Override
	protected void doGet(HttpServletRequest req, HttpServletResponse resp) {
		Object obj = get(req, resp);
		render(obj, req, resp);
	}

	@Override
	protected void doPost(HttpServletRequest req, HttpServletResponse resp) {
		Object obj = post(req, resp);
		render(obj, req, resp);
	}

	@Override
	protected void doPut(HttpServletRequest req, HttpServletResponse resp) {
		Object obj = put(req, resp);
		render(obj, req, resp);
	}

	@Override
	protected void doDelete(HttpServletRequest req, HttpServletResponse resp) {
		Object obj = delete(req, resp);
		render(obj, req, resp);
	}

	public Object get(HttpServletRequest req, HttpServletResponse resp) {
		return null;
	};

	public Object post(HttpServletRequest req, HttpServletResponse resp) {
		return null;
	};

	public Object put(HttpServletRequest req, HttpServletResponse resp) {
		return null;
	};

	public Object delete(HttpServletRequest req, HttpServletResponse resp) {
		return null;
	};
}

File: src/main/kotlin/com/haozileung/infra/web/BaseServlet.kt
package com.haozileung.infra.web;

import com.alibaba.fastjson.JSON
import com.google.common.base.Strings
import com.google.inject.Injector
import org.beetl.ext.servlet.ServletGroupTemplate
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.IOException
import java.io.PrintWriter
import javax.servlet.http.HttpServlet
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

abstract class BaseServlet : HttpServlet() {
    private final val logger: Logger = LoggerFactory.getLogger(BaseServlet::class.java);

    override fun init() {
        (Initializer.instance as Injector).injectMembers(this);
    }

    fun renderView(code: Int, view: String, req: HttpServletRequest, resp: HttpServletResponse) {
        resp.contentType = "text/html;charset=utf-8";
        resp.status = code;
        if (!Strings.isNullOrEmpty(view)) {
            try {
                ServletGroupTemplate.instance().render(view, req, resp);
            } catch (e: Exception) {
                logger.error(e.message, e);
                try {
                    resp.sendError(HttpServletResponse.SC_NOT_FOUND);
                } catch (e1: IOException) {
                    logger.error(e1.message, e1);
                }
            }
        }
    }

    fun renderJSON(code: Int, data: Any?, resp: HttpServletResponse) {
        resp.contentType = "application/json;charset=utf-8";
        resp.status = code;
        try {
            var pr: PrintWriter = resp.writer;
            pr.print(JSON.toJSONString(data));
            pr.flush();
        } catch (e: IOException) {
            logger.error(e.message, e);
        }
    }

    fun render(result: Any?, req: HttpServletRequest, resp: HttpServletResponse) {
        if (result != null) {
            if (result is String) {
                if (result.startsWith("redirect:")) {
                    val v = result.replace("redirect:", "");
                    try {
                        resp.sendRedirect(v);
                    } catch (e: IOException) {
                        logger.error(e.message, e);
                    }
                }
                renderView(resp.status, result, req, resp);
            } else {
                renderJSON(resp.status, result, resp);
            }
        }
    }

    override fun doGet(req: HttpServletRequest, resp: HttpServletResponse) {
        val obj: Any = get(req, resp);
        render(obj, req, resp);
    }


    override fun doPost(req: HttpServletRequest, resp: HttpServletResponse) {
        val obj: Any = get(req, resp);
        render(obj, req, resp);
    }

    override fun doPut(req: HttpServletRequest, resp: HttpServletResponse) {
        val obj: Any = get(req, resp);
        render(obj, req, resp);
    }

    override fun doDelete(req: HttpServletRequest, resp: HttpServletResponse) {
        val obj: Any = get(req, resp);
        render(obj, req, resp);
    }

    open fun get(req: HttpServletRequest, resp: HttpServletResponse): Any {
        return Any();
    }

    open fun post(req: HttpServletRequest, resp: HttpServletResponse): Any {
        return Any();
    }

    open fun put(req: HttpServletRequest, resp: HttpServletResponse): Any {
        return Any();
    }

    open fun delete(req: HttpServletRequest, resp: HttpServletResponse): Any {
        return Any();
    }

}